pkg/miniogw/gateway-storj.go

```diff
@@ -5,27 +5,196 @@ package storj
 
 import (
 	"context"
+	"crypto/sha256"
+	"flag"
+	"fmt"
 	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
 	"time"
 
 	"github.com/minio/cli"
+	"github.com/vivint/infectious"
+	"storj.io/storj/pkg/eestream"
 
-	"github.com/minio/minio/cmd"
+	minio "github.com/minio/minio/cmd"
 	"github.com/minio/minio/pkg/auth"
 	"github.com/minio/minio/pkg/hash"
 )
 
+var (
+	pieceBlockSize = flag.Int("piece_block_size", 4*1024, "block size of pieces")
+	key            = flag.String("key", "a key", "the secret key")
+	rsk            = flag.Int("required", 20, "rs required")
+	rsn            = flag.Int("total", 40, "rs total")
+)
+
+// S3CliAPI contains parameters for accessing the Storj network
+type S3CliAPI struct {
+	totalBuckets int
+	bucketlist   []S3BucketList
+}
+
+//S3BucketList structure
+type S3BucketList struct {
+	bucket   minio.BucketInfo
+	filelist S3FileList
+}
+
+//S3FileList structure
+type S3FileList struct {
+	totalFiles int
+	file       minio.ListObjectsInfo
+}
+
+// gS3Gateway global S3 interface structure
+var gS3CliAPI S3CliAPI
+
+//createDummyBucketList function initializes sample buckets and files in each bucket
+func createDummyBucketList() {
+	gS3CliAPI.bucketlist = gS3CliAPI.bucketlist[:0]
+	for i := 0x00; i < 0x0A; i++ {
+		gS3CliAPI.bucketlist = append(gS3CliAPI.bucketlist,
+			S3BucketList{
+				bucket: minio.BucketInfo{
+					Name:    "TestBucket" + strconv.Itoa(i+1),
+					Created: time.Now(),
+				},
+			},
+		)
+	} /* end of for loop */
+
+	for i := 0x00; i < 0x0A; i++ {
+		gS3CliAPI.bucketlist[i].filelist.file.IsTruncated = false
+		gS3CliAPI.bucketlist[i].filelist.file.Objects = gS3CliAPI.bucketlist[i].filelist.file.Objects[:0]
+		for j := 0x00; j < 0x0A; j++ {
+			gS3CliAPI.bucketlist[i].filelist.file.Objects = append(
+				gS3CliAPI.bucketlist[i].filelist.file.Objects,
+				minio.ObjectInfo{
+					Bucket:      gS3CliAPI.bucketlist[i].bucket.Name,
+					Name:        gS3CliAPI.bucketlist[i].bucket.Name + "file" + strconv.Itoa(j+1),
+					ModTime:     time.Now(),
+					Size:        int64(100 + i),
+					ContentType: "application/octet-stream",
+				},
+			)
+		}
+	}
+	gS3CliAPI.totalBuckets = len(gS3CliAPI.bucketlist)
+	fmt.Println("bucket name = ", gS3CliAPI.bucketlist)
+}
+
 func init() {
-	cmd.RegisterGatewayCommand(cli.Command{
+	minio.RegisterGatewayCommand(cli.Command{
 		Name:            "storj",
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
+
+	// create dummy bucket list
+	createDummyBucketList()
```

Comment:
instead of having this function mutate global state, you can have the Action on line 94 have a "closure" which is a function that is lexically bound to some specific state.

for instance, this could be:

```
func init() {
   s := &Storj{}
   s.createDummyBucketList()
   minio.RegisterGatewayCommand(cli.Command{
     Name: "storj",
     Usage: "Storj",
     Action: func(ctx *cli.Context) {
        minio.StartGateway(ctx, s)
     },
     HideHelpCommand: true,
   })
}
```

notice how `createDummyBucketList` is defined on a specific struct, thus not needing a global.

Reply:
Modified as suggested to eliminate global references